kubechain/internal/controller/llm/llm_controller.go

```diff
@@ -43,55 +49,262 @@ type LLMReconciler struct {
+			if config.Timeout != nil {
+				timeoutDuration := time.Duration(*config.Timeout) * time.Second
+				providerOpts = append(providerOpts, mistral.WithTimeout(timeoutDuration))
+			}
+
+			// Set RandomSeed if provided
+			if config.RandomSeed != nil {
+				commonOpts = append(commonOpts, llms.WithSeed(*config.RandomSeed))
+			}
+		}
+
+		// Create the Mistral model with the provider options
+		model, err = mistral.New(providerOpts...)
+
+		// Pass any common options to the model during generation test
+		if len(commonOpts) > 0 {
+			commonOpts = append(commonOpts, llms.WithMaxTokens(1), llms.WithTemperature(0))
+			_, err = llms.GenerateFromSinglePrompt(ctx, model, "test", commonOpts...)
+			if err != nil {
+				return fmt.Errorf("mistral validation failed with options: %w", err)
+			}
+			return nil
+		}
+
+	case "google":
+		if llm.Spec.APIKeyFrom == nil {
+			return fmt.Errorf("apiKeyFrom is required for google")
+		}
+		providerOpts := []googleai.Option{googleai.WithAPIKey(apiKey)}
+		if llm.Spec.ProviderConfig.GoogleConfig != nil {
+			if llm.Spec.ProviderConfig.GoogleConfig.CloudProject != "" {
+				providerOpts = append(providerOpts, googleai.WithCloudProject(llm.Spec.ProviderConfig.GoogleConfig.CloudProject))
+			}
+			if llm.Spec.ProviderConfig.GoogleConfig.CloudLocation != "" {
+				providerOpts = append(providerOpts, googleai.WithCloudLocation(llm.Spec.ProviderConfig.GoogleConfig.CloudLocation))
+			}
+		}
+		if llm.Spec.BaseConfig.Model != "" {
+			providerOpts = append(providerOpts, googleai.WithDefaultModel(llm.Spec.BaseConfig.Model))
+		}
+		model, err = googleai.New(ctx, providerOpts...)
+
+	case "vertex":
+		if llm.Spec.ProviderConfig.VertexConfig == nil {
+			return fmt.Errorf("vertexConfig is required for vertex")
+		}
+		config := llm.Spec.ProviderConfig.VertexConfig
+		providerOpts := []googleai.Option{
+			googleai.WithCloudProject(config.CloudProject),
+			googleai.WithCloudLocation(config.CloudLocation),
+		}
+		if llm.Spec.APIKeyFrom != nil && apiKey != "" {
+			providerOpts = append(providerOpts, googleai.WithCredentialsJSON([]byte(apiKey)))
+		}
+		if llm.Spec.BaseConfig.Model != "" {
+			providerOpts = append(providerOpts, googleai.WithDefaultModel(llm.Spec.BaseConfig.Model))
+		}
+		model, err = vertex.New(ctx, providerOpts...)
+
+	default:
+		return fmt.Errorf("unsupported provider: %s", llm.Spec.Provider)
+	}
 
-	client := &http.Client{}
-	resp, err := client.Do(req)
 	if err != nil {
-		return fmt.Errorf("failed to make request: %w", err)
+		return fmt.Errorf("failed to initialize %s client: %w", llm.Spec.Provider, err)
 	}
-	defer func() {
-		if err := resp.Body.Close(); err != nil {
-			fmt.Printf("Error closing response body: %v\n", err)
-		}
-	}()
 
-	if resp.StatusCode != http.StatusOK {
-		return fmt.Errorf("invalid API key (status code: %d)", resp.StatusCode)
+	// Validate with a test call
+	_, err = llms.GenerateFromSinglePrompt(ctx, model, "test", llms.WithTemperature(0), llms.WithMaxTokens(1))
```

Comment:
this makes sense as a test call. I wonder if the providers offer other more dedicated testing stuff. fine for now though